api/src/routes/catalogs.rs

```diff
@@ -0,0 +1,522 @@
+        Err(e) => {
+            return response::internal_error(&format!("Error getting catalog info: {:?}", e));
+        }
+    };
+    // Catalogs can't be part of the protected names and can't start with "system."
+    let collection_names = collection_names
+        .into_iter()
+        .filter(|name| {
+            !PROTECTED_COLLECTION_NAMES.contains(&name.as_str()) && !name.starts_with("system.")
+        })
+        .collect::<Vec<String>>();
+    // Remove the prefix to get the catalog names
+    let mut catalog_names: Vec<String> = collection_names
+        .iter()
+        .map(|name| name.trim_start_matches("catalog_").to_string())
+        .collect();
+    catalog_names.sort();
+    let mut catalogs = Vec::new();
+    if params.get_details.unwrap_or_default() {
+        for catalog in catalog_names {
+            match db
+                .run_command(doc! {
+                    "collstats": get_collection_name(&catalog)
+                })
+                .await
+            {
+                Ok(d) => catalogs.push(doc! {"name": catalog, "details": d}),
+                Err(e) => {
+                    return response::internal_error(&format!(
+                        "Error getting catalog info: {:?}",
+                        e
+                    ));
+                }
+            };
+        }
+    } else {
+        // If no details requested, just return the names
+        for catalog in catalog_names {
+            catalogs.push(doc! { "name": catalog });
+        }
+    }
+    // Serialize catalogs
+    match serde_json::to_value(&catalogs) {
+        Ok(v) => return response::ok("success", v),
+        Err(e) => {
+            return response::internal_error(&format!("Error serializing catalog info: {:?}", e));
+        }
+    };
+}
+
+#[derive(serde::Deserialize, Clone, ToSchema)]
+struct CountQuery {
+    filter: Option<serde_json::Value>,
+}
+
+/// Run a count query on a catalog
+#[utoipa::path(
+    post,
+    path = "/catalogs/{catalog_name}/queries/count",
+    params(
+        ("catalog_name" = String, Path, description = "Name of the catalog (case insensitive), e.g., 'ztf'"),
+    ),
+    request_body = CountQuery,
+    responses(
+        (status = 200, description = "Count of documents in the catalog", body = serde_json::Value),
+        (status = 400, description = "Bad request"),
+        (status = 500, description = "Internal server error")
+    )
+)]
+#[post("/catalogs/{catalog_name}/queries/count")]
+pub async fn post_catalog_count_query(
+    db: web::Data<Database>,
+    catalog_name: web::Path<String>,
+    web::Json(query): web::Json<CountQuery>,
+) -> HttpResponse {
+    // Validate catalog name
+    if catalog_name.is_empty() {
+        return response::bad_request("Catalog name cannot be empty");
+    }
+    let catalog_name = catalog_name.into_inner();
+    // Check that there is a collection with this catalog name
+    let collection_names = match db.list_collection_names().await {
```

Comment:
Is this necessary? It seems like getting a document count shouldn't involve two queries.

Reply:
I would like to return a 404 if a user requests data from a nonexistent catalog, rather than an empty response. I will look for a way to do that in a single call though.

Reply:
I forget exactly how this works, but the mongo crate is generally lazy so it should be possible to run a query on an arbitrary collection name, and you'll get an error if the collection doesn't exist. You should be able to inspect the error code to determine whether to return 404 or 500.

Reply:
lmao, yeah no the mongodb crate - and they explain that in their docs - made the decision to NOT do any db round trips when defining `collection`s and such, which I do agree with actually. But yeah in mongo land I think that querying a DB that doesn't exist just works.

Reply:
![image](https://github.com/user-attachments/assets/ec495835-351c-4e7f-a7d7-40b78f458e53)